.github/workflows/metal-cpp-golang-rust.yml

```diff
@@ -0,0 +1,521 @@
+        path: ./icicle/backend/metal
+        ssh-key: ${{ secrets.METAL_PULL_KEY }}
+        ref: ${{ needs.extract-metal-backend-branch.outputs.metal-backend-branch }}
+    - name: Get METAL Backend Commit SHA
+      if: needs.check-changed-files.outputs.golang == 'true' || needs.check-changed-files.outputs.cpp == 'true'
+      working-directory: ./icicle/backend/metal
+      id: extract-metal-sha
+      run: |
+        METAL_BACKEND_SHA=$(git rev-parse HEAD)
+        echo "METAL Backend Commit SHA: $METAL_BACKEND_SHA"
+        echo "metal-backend-sha=$METAL_BACKEND_SHA" >> $GITHUB_OUTPUT
+    - name: Set METAL backend flag
+      if: needs.check-changed-files.outputs.golang == 'true' || needs.check-changed-files.outputs.cpp == 'true'
+      id: metal-flag
+      run: |
+        METAL_BACKEND_SHA=${{ steps.extract-metal-sha.outputs.metal-backend-sha }}
+        COMMIT_FILE="gh_commit_sha_hash_${METAL_BACKEND_SHA}"
+        if [ "${{ needs.extract-metal-backend-branch.outputs.metal-backend-branch }}" == "main" ]; then
+          INSTALL_PATH=${{ github.workspace }}/../../main_lib/rust-examples
+          echo "INSTALL_PATH=${{ github.workspace }}/../../main_lib/rust-examples" >> $GITHUB_OUTPUT
+          COMMIT_FILE_PATH=${INSTALL_PATH}/lib/${COMMIT_FILE}
+          echo "COMMIT_FILE_PATH=${INSTALL_PATH}/lib/${COMMIT_FILE}" >> $GITHUB_OUTPUT
+
+          if [ -f "$COMMIT_FILE_PATH" ]; then
+            echo "${COMMIT_FILE} exists in ${INSTALL_PATH}/lib. Setting METAL_FLAG="
+            echo "METAL_FLAG=" >> $GITHUB_OUTPUT
+          else
+            echo "${COMMIT_FILE} does not exist in ${INSTALL_PATH}/lib. Setting METAL_FLAG=-DMETAL_BACKEND=local"
+            echo "METAL_FLAG=-DMETAL_BACKEND=local" >> $GITHUB_OUTPUT
+            rm -rf ${INSTALL_PATH}
+          fi
+          echo "CMAKE_INSTALL_PREFIX=-DCMAKE_INSTALL_PREFIX=${INSTALL_PATH}" >> $GITHUB_OUTPUT
+          echo "ICICLE_BACKEND_INSTALL_DIR=${INSTALL_PATH}/lib" >> $GITHUB_OUTPUT
+        else
+          INSTALL_PATH=${{ github.workspace }}/../../temp_lib/rust-examples
+          echo "INSTALL_PATH=${{ github.workspace }}/../../temp_lib/rust-examples" >> $GITHUB_OUTPUT
+          COMMIT_FILE_PATH=${INSTALL_PATH}/lib/${COMMIT_FILE}
+          echo "COMMIT_FILE_PATH=${INSTALL_PATH}/lib/${COMMIT_FILE}" >> $GITHUB_OUTPUT
+          if [ -f "$COMMIT_FILE_PATH" ]; then
+            echo "${COMMIT_FILE} exists in ${INSTALL_PATH}/lib. Setting METAL_FLAG="
+            echo "METAL_FLAG=" >> $GITHUB_OUTPUT
+          else
+            echo "${COMMIT_FILE} does not exist in ${INSTALL_PATH}/lib. Setting METAL_FLAG=-DMETAL_BACKEND=local"
+            echo "METAL_FLAG=-DMETAL_BACKEND=local" >> $GITHUB_OUTPUT
+            rm -rf ${INSTALL_PATH}
+          fi
+          echo "CMAKE_INSTALL_PREFIX=-DCMAKE_INSTALL_PREFIX=${INSTALL_PATH}" >> $GITHUB_OUTPUT
+          echo "ICICLE_BACKEND_INSTALL_DIR=${INSTALL_PATH}/lib" >> $GITHUB_OUTPUT
+        fi
```

Comment:
I would move this to a reusable workflow. No reason to rewrite (copy pasta) this step on each job.
It can also be backend and feature (hash, examples, curve, field, etc) agnostic and accept inputs that dictate which backend and what paths to check or build to